### PR TITLE
Fix type initialization ordering under R2R

### LINKED ISSUE #84007: Test failure Loader\\classloader\\TypeInitialization\\CctorsWithSideEffects\\CctorForWrite\\CctorForWrite.cmd
Run: [runtime-coreclr outerloop 20230327.2](https://dev.azure.com/dnceng-public/public/_build/results?buildId=218637&view=ms.vss-test-web.build-test-results-tab&runId=4093860&resultId=116839&paneView=debug)

Failed test:
```

R2R-CG2 windows arm64 Checked no_tiered_compilation @ Windows.11.Arm64.Open

- Loader\\classloader\\TypeInitialization\\CctorsWithSideEffects\\CctorForWrite\\CctorForWrite.cmd

R2R-CG2 linux x64 Checked no_tiered_compilation @ (Alpine.315.Amd64.Open)Ubuntu.1804.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.15-helix-amd64

- Loader/classloader/TypeInitialization/CctorsWithSideEffects/CctorForWrite/CctorForWrite.sh

R2R-CG2 windows x86 Checked no_tiered_compilation @ Windows.10.Amd64.Open

- Loader\\classloader\\TypeInitialization\\CctorsWithSideEffects\\CctorForWrite\\CctorForWrite.cmd

R2R-CG2 windows x64 Checked @ Windows.10.Amd64.Open

- Loader\\classloader\\TypeInitialization\\CctorsWithSideEffects\\CctorForWrite\\CctorForWrite.cmd

R2R-CG2 windows x64 Checked @ Windows.10.Amd64.Open

- Loader\\classloader\\TypeInitialization\\CctorsWithSideEffects\\CctorForWrite\\CctorForWrite.cmd
... ...
```
**Error message:**
```
Return code:      1
Raw output file:      C:\h\w\AA970943\w\AFD009BD\uploads\Reports\Loader.classloader\TypeInitialization\CctorsWithSideEffects\CctorForWrite\CctorForWrite.output.txt
Raw output:
BEGIN EXECUTION
CctorForWrite.dll
1 file(s) copied.
20:30:54.97
Response file: C:\h\w\AA970943\w\AFD009BD\e\Loader\classloader\TypeInitialization\CctorsWithSideEffects\CctorForWrite\CctorForWrite.dll.rsp
C:\h\w\AA970943\w\AFD009BD\e\Loader\classloader\TypeInitialization\CctorsWithSideEffects\CctorForWrite\IL-CG2\CctorForWrite.dll
-o:C:\h\w\AA970943\w\AFD009BD\e\Loader\classloader\TypeInitialization\CctorsWithSideEffects\CctorForWrite\CctorForWrite.dll
--targetarch:arm64
--targetos:windows
--verify-type-and-field-layout
--method-layout:random
-r:C:\h\w\AA970943\p\System..dll
-r:C:\h\w\AA970943\p\Microsoft..dll
-r:C:\h\w\AA970943\p\xunit.*.dll
-r:C:\h\w\AA970943\p\mscorlib.dll
-r:C:\h\w\AA970943\p\netstandard.dll
-O
" "dotnet" "C:\h\w\AA970943\p\crossgen2\crossgen2.dll" @"C:\h\w\AA970943\w\AFD009BD\e\Loader\classloader\TypeInitialization\CctorsWithSideEffects\CctorForWrite\CctorForWrite.dll.rsp"   -r:C:\h\w\AA970943\w\AFD009BD\e\Loader\classloader\TypeInitialization\CctorsWithSideEffects\CctorForWrite\IL-CG2*.dll"
Emitting R2R PE file: C:\h\w\AA970943\w\AFD009BD\e\Loader\classloader\TypeInitialization\CctorsWithSideEffects\CctorForWrite\CctorForWrite.dll
" "dotnet" "C:\h\w\AA970943\p\r2rdump\r2rdump.dll" --header --sc --in C:\h\w\AA970943\w\AFD009BD\e\Loader\classloader\TypeInitialization\CctorsWithSideEffects\CctorForWrite\CctorForWrite.dll --out C:\h\w\AA970943\w\AFD009BD\e\Loader\classloader\TypeInitialization\CctorsWithSideEffects\CctorForWrite\CctorForWrite.dll.r2rdump --val"
20:30:57.25
"C:\h\w\AA970943\p\corerun.exe" -p "System.Reflection.Metadata.MetadataUpdater.IsSupported=false"  CctorForWrite.dll
Expected: 100
Actual: 1
END EXECUTION - FAILED
FAILED
Test failed. Trying to see if dump file was created in C:\cores since 3/27/2023 8:30:54 PM
Test Harness Exitcode is : 1
To run the test:

set CORE_ROOT=C:\h\w\AA970943\p
C:\h\w\AA970943\w\AFD009BD\e\Loader\classloader\TypeInitialization\CctorsWithSideEffects\CctorForWrite\CctorForWrite.cmd
Expected: True
Actual:   False


Stack trace
   at Loader_classloader._TypeInitialization_CctorsWithSideEffects_CctorForWrite_CctorForWrite_._TypeInitialization_CctorsWithSideEffects_CctorForWrite_CctorForWrite_cmd()
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

### PR DESCRIPTION
**Pull request by markples**
Expand the fix in #80485 to cover helpers including common ones for R2R as well as some generic helpers for JIT.

This would cause regressions (100ks code size in diffs) if done alone, so it also includes checks on whether a helper may trigger a cctor and whether it is beforefieldinit. The precise ordering of the fix is only required for a non-beforefieldinit ctor. This recovers the (small) losses from #80485 and mitigates this fix.

As before, the primary diff is moving type initialization into (for correctness) or out of (the mitigation optimization) a JIT tree. These leads to secondary effects as values may need to be preserved around the type initialization code.

The end result of the fix+mitigations is very little code size change, but viewing the diffs suggests possible future code improvements. However, the impact of any of these should be established first.
- A type initializer will never invoke itself
- When type initialization is removed/CSEed late, it is too late for forward subst/etc., so the loss of the intact tree can still impact code generation
- A custom calling convention for the static helpers (tuned to the common case of no type initialization needing to occur) could help with the impact of calling it in the middle of a calculation
- Code can use beforefieldinit in more place

Fixes #84007

### CODE REVIEW

**Comment by markples** [comment]
Some code size measurements:

R2R - [ReadyToRun fix](https://dev.azure.com/dnceng-public/public/_build/results?buildId=230883&view=ms.vss-build-web.run-extensions-tab)
R2R+old - [ReadyToRun fix + applying opt to previous fix](https://dev.azure.com/dnceng-public/public/_build/results?buildId=231224&view=ms.vss-build-web.run-extensions-tab)
undo old - [Simply undoing the old fix (seeing its cost)](https://dev.azure.com/dnceng-public/public/_build/results?buildId=232145&view=ms.vss-build-web.run-extensions-tab)
old - [Just applying opt to previous fix](https://dev.azure.com/dnceng-public/public/_build/results?buildId=233091&view=ms.vss-build-web.run-extensions-tab)
R2R+old 2 - [Redoing R2R+old after factoring](https://dev.azure.com/dnceng-public/public/_build/results?buildId=236681&view=results)

Minor variations appear to be due to the repo changing day-to-day, so the takeaway is simply that applying the optimization to the old fix gets that regression back and the new fix+optimization has very little impact.  #N/A occurred when the job didn't find the .mch file.

host | target | opt | mch | R2R | R2R+old | undo old | old | R2R+old 2
-- | -- | -- | -- | --: | --: | --: | --: | --:
x64 | linux arm64 | min | coreclr_tests.run| -8 | -7200 | -7264 | -7192 | -7200
  |   |   | libraries_tests.pmi| 12 | 12 |   |   | 12
  |   | full | benchmarks.run|   | -48 | -48 | -48 | -48
  |   |   | coreclr_tests.run|   | -728 | -732 | -664 | -672
  |   |   | libraries.crossgen2| -72 | -72 |   |   | -100
  |   |   | libraries.pmi|   | -24 | -28 | -24 | -24
  |   |   | libraries_tests.pmi| 0 | -544 | -544 | -540 | -544
  | osx arm64 | min | coreclr_tests.run| -8 | -7200 | -7264 | -7192 | -7200
  |   |   | libraries_tests.pmi| 12 | 12 |   |   | 12
  |   | full | benchmarks.run|   | -56 | -56 | -56 | -56
  |   |   | coreclr_tests.run|   | -728 | -732 | -636 | -636
  |   |   | libraries.crossgen2| -140 | -140 |   |   | -140
  |   |   | libraries.pmi|   | -28 | -32 | -28 | -28
  |   |   | libraries_tests.pmi|   | -536 | -536 | #N/A | -532
  | linux x64 | min | coreclr_tests.run| 12 | -10456 | -10552 | -10747 | -10735
  |   |   | libraries_tests.pmi| 14 | 14 |   |   | 14
  |   | full | benchmarks.run|   | -42 | -42 | -48 | -48
  |   |   | coreclr_tests.run|   | -1257 | -1260 | -1254 | -1262
  |   |   | libraries.crossgen2| -159 | -159 |   |   | -195
  |   |   | libraries.pmi|   | -45 | -48 | -44 | -44
  |   |   | libraries_tests.pmi|   | -395 | -395 | -390 | -402
  | win arm64 | min | coreclr_tests.run| -8 | -7200 | -7264 | -7192 | -7200
  |   |   | libraries_tests.pmi| 12 | 12 |   |   | 12
  |   | full | benchmarks.run|   | -48 | -48 | -48 | -48
  |   |   | coreclr_tests.run|   | -716 | -720 | -668 | -668
  |   |   | libraries.crossgen2| -140 | -140 |   |   | -128
  |   |   | libraries.pmi|   | -56 | -60 | -56 | -56
  |   |   | libraries_tests.pmi|   | -548 | -548 | -548 | -548
  | win x64 | min | aspnet.run|   | -160 | -160 | #N/A | -160
  |   |   | coreclr_tests.run| 12 | -10436 | -10532 | -10755 | -10743
  |   |   | libraries_tests.pmi| 13 | 13 |   |   | 13
  |   | full | aspnet.run|   | 2 | 2 | #N/A | 1
  |   |   | benchmarks.run|   | -36 | -36 | -36 | -36
  |   |   | coreclr_tests.run|   | -1231 | -1243 | -1221 | -1233
  |   |   | libraries.crossgen2| -252 | -252 |   |   | -252
  |   |   | libraries.pmi|   | -112 | -115 | -108 | -108
  |   |   | libraries_tests.pmi|   | -470 | -470 | -454 | -452
x86 | linux arm | min | coreclr_tests.run| -6 | -5346 | -5344 | -5340 | -5346
  |   |   | libraries_tests.pmi| 6 | 6 |   |   | 6
  |   | full | benchmarks.run|   | 24 | 24 | 10 | 10
  |   |   | coreclr_tests.run|   | -236 | -238 | -232 | -236
  |   |   | libraries.crossgen2| -170 | -170 |   |   | -62
  |   |   | libraries.pmi|   | 188 | 196 | 182 | 182
  |   |   | libraries_tests.pmi|   | -120 | -120 | -126 | -126
  | win x86 | min | coreclr_tests.run| 2 | -8533 | -8582 | -8535 | -8533
  |   |   | libraries_tests.pmi| 9 | 9 |   |   | 9
  |   | full | benchmarks.run|   | -52 | -52 | -62 | -62
  |   |   | coreclr_tests.run|   | -1623 | -1631 | -1627 | -1630
  |   |   | libraries.crossgen2| -194 | -194 |   |   | -172
  |   |   | libraries.pmi|   | -99 | -101 | -102 | -102
  |   |   | libraries_tests.pmi|   | -225 | -225 | -201 | -201

**Comment by markples** [comment]
test results:

runtime
- jit-format - #84780
- InstrumentedTiers - Nuget-Migrations

outerloop
- CoreCLR Product Build linux (and linux-musl) arm64 debug - #84802
- jit-format - #84780
- slowpatheltenter/leave - #84750
- crossgen2determinism, coreroot_determinism, BasicTestWithMcj, crossgen2smoke_donotalwaysusecrossgen2 - #84752
- baseservices/threading/generic/* - ?
- Frozen - #84462
- tlstest - ?

extra-platforms has tons of failures. It's very hard to tell what's going on, but other jobs that I've looked at seem to have similar failures.

**Comment by markples** [comment]
@AndyAyersMS Could you please review this one?

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -4097,7 +4097,8 @@ GenTree* Compiler::impImportCnsTreeFromBuffer(uint8_t* buffer, var_types valueTy
 GenTree* Compiler::impImportStaticFieldAccess(CORINFO_RESOLVED_TOKEN* pResolvedToken,
                                               CORINFO_ACCESS_FLAGS    access,
                                               CORINFO_FIELD_INFO*     pFieldInfo,
-                                              var_types               lclTyp)
+                                              var_types               lclTyp,
+                                              /* OUT */ bool*         pIsHoistable)
```
Comment (AndyAyersMS) [inline src/coreclr/jit/importer.cpp:4101]: Add a header comment? I know we still have many without these comments but I try to add one for the methods I'm changing.

**Review by AndyAyersMS — APPROVED** [review]
LGTM with one commenting nit.

File: src/coreclr/jit/importer.cpp (hunk 1 above)
Comment (markples) [reply src/coreclr/jit/importer.cpp:4128]: I attempted one.. learning a bit about those types in the process.

(the merge conflict was between code that I moved and main removed)

**Pull request merged**